### PR TITLE
dialects: add linalg.fill 

### PR DESCRIPTION
**Pull request by kayode-gif**

### CODE REVIEW

File: xdsl/dialects/linalg.py [hunk 1]
```diff
@@ -416,12 +416,55 @@ def __init__(
...
+        else:
+            result_types = res
+
+        super().__init__(
+            operands=(inputs, outputs),
+            result_types=result_types,
+        )
+        # Numeric casting is performed on the value operand, promoting it to the same data type as the output.
+        if outputs:
+            self.inputs = [cast(outputs[0].type, operand) for operand in inputs]
```
Comment (superlopuh) [inline xdsl/dialects/linalg.py:458]: this is not needed, the casting is at runtime

Comment (superlopuh) [reply xdsl/dialects/linalg.py:458]: The meaning is that you can fill a tensor of f64 with inputs of type f32, these types don't need to match. I don't know what kind of validation MLIR does on compatibility of types.

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
As you can see from the CI output, the fill operation in MLIR expects a scalar input, not a tensor. We might want to implement the same verification. The best thing to do would be to look at the MLIR source code to find the verifier for the fill op

File: xdsl/dialects/linalg.py (hunk 1 above)
Comment (kayode-gif) [reply xdsl/dialects/linalg.py:458]: ahh that makes sense, thanks

File: tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/invalid_ops.mlir [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// RUN: xdsl-opt %s | xdsl-opt | mlir-opt --allow-unregistered-dialect | filecheck %s
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/invalid_ops.mlir:1]:
```suggestion
// RUN: xdsl-opt --verify-diagnostics %s | filecheck %s
```
This is to test our own verifier, right? Not mlir-opt

File: tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir [hunk 3]
```diff
@@ -38,9 +45,11 @@ linalg.generic {indexing_maps = [affine_map<(d0, d1) -> ()>, affine_map<(d0, d1)
...
 // CHECK-NEXT:    %1:2 = "test.op"() : () -> (tensor<2x3xf32>, tensor<2x3xf32>)
 // CHECK-NEXT:    %2 = linalg.add ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#1 : tensor<2x3xf32>) -> tensor<2x3xf32>
-// CHECK-NEXT:    %3 = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#0 : tensor<2x3xf32>) {
+// CHECK-NEXT:    %cst = arith.constant 0.000000e+00 : f32
+// CHECK-NEXT:    %3 = linalg.fill ins(%cst : f32) outs(%1#0 : tensor<2x3xf32>) -> tensor<2x3xf32>
+// CHECK-NEXT:    %4 = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#0 : tensor<2x3xf32>) {
 // CHECK-NEXT:    ^bb0(%in: f32, %in_0: f32, %out: f32):
-// CHECK-NEXT:      %4 = arith.addf %in, %in_0 : f32
-// CHECK-NEXT:      linalg.yield %4 : f32
+// CHECK-NEXT:      %5 = arith.addf %in, %in_0 : f32
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir:52]: this sort of diff is why I usually append to the end of the test file

Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir:52]:
```suggestion
// CHECK-NEXT:      %{{.*}} = arith.addf %in, %in_0 : f32
```
But if we want the generic to be at the end of the test file then let's fix this now

File: tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir [hunk 4]
```diff
@@ -38,9 +45,11 @@ linalg.generic {indexing_maps = [affine_map<(d0, d1) -> ()>, affine_map<(d0, d1)
 // CHECK-NEXT:    }
 // CHECK-NEXT:    %1:2 = "test.op"() : () -> (tensor<2x3xf32>, tensor<2x3xf32>)
 // CHECK-NEXT:    %2 = linalg.add ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#1 : tensor<2x3xf32>) -> tensor<2x3xf32>
-// CHECK-NEXT:    %3 = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#0 : tensor<2x3xf32>) {
+// CHECK-NEXT:    %cst = arith.constant 0.000000e+00 : f32
+// CHECK-NEXT:    %3 = linalg.fill ins(%cst : f32) outs(%1#0 : tensor<2x3xf32>) -> tensor<2x3xf32>
+// CHECK-NEXT:    %4 = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#0 : tensor<2x3xf32>) {
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir:50]: Sorry, this one also would be great
```suggestion
// CHECK-NEXT:    %{{.*}} = linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel"]} ins(%1#0, %1#0 : tensor<2x3xf32>, tensor<2x3xf32>) outs(%1#0 : tensor<2x3xf32>) {
```

Comment (superlopuh) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/linalg/ops.mlir:50]: That way we eliminate future spurious diffs

**Comment by gabrielrodcanal** [comment]
Haven't really checked the custom format in the operation definition in MLIR, but I've found an example in which `-> type($res)` is optional (I'm referring to https://github.com/xdslproject/xdsl/blob/ae2054d727e261e8873115e182d8bb53bc7ca0f9/xdsl/dialects/linalg.py#L439). See the example here: https://github.com/Xilinx/mlir-air/blob/82d23361cb3843e5d8fc01d746d513f25a224125/test/airhost/51_air_mmult_2x2_channel_broadcast/air.mlir#L30

I'm not sure if this was a mistake or it's just provisional until the cutom format API supports optional groups, but just in case I thought I'd have you know!

**Comment by superlopuh** [comment]
Oh interesting, should be easy to fix, can you please file an issue with this info?

**Comment by gabrielrodcanal** [comment]
See #2228.

**Pull request merged**